File: go/test/berufsbildung_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/berufsbildung-daten-exporte-sdk/go"
	"github.com/voxgig-sdk/berufsbildung-daten-exporte-sdk/go/core"

	vs "github.com/voxgig-sdk/berufsbildung-daten-exporte-sdk/go/utility/struct"
)

func TestBerufsbildungEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Berufsbildung(nil)
		if ent == nil {
			t.Fatal("expected non-nil BerufsbildungEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := berufsbildungBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "berufsbildung." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set BERUFSBILDUNGDATENEXPORTE_TEST_BERUFSBILDUNG_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		berufsbildungRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.berufsbildung", setup.data)))
		var berufsbildungRef01Data map[string]any
		if len(berufsbildungRef01DataRaw) > 0 {
			berufsbildungRef01Data = core.ToMapAny(berufsbildungRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = berufsbildungRef01Data

		// LIST
		berufsbildungRef01Ent := client.Berufsbildung(nil)
		berufsbildungRef01Match := map[string]any{}

		berufsbildungRef01ListResult, err := berufsbildungRef01Ent.List(berufsbildungRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, berufsbildungRef01ListOk := berufsbildungRef01ListResult.([]any)
		if !berufsbildungRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", berufsbildungRef01ListResult)
		}

		// LOAD
		berufsbildungRef01MatchDt0 := map[string]any{}
		berufsbildungRef01DataDt0Loaded, err := berufsbildungRef01Ent.Load(berufsbildungRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if berufsbildungRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func berufsbildungBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "berufsbildung", "BerufsbildungTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read berufsbildung test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse berufsbildung test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"berufsbildung01", "berufsbildung02", "berufsbildung03", "export01", "export02", "export03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("BERUFSBILDUNGDATENEXPORTE_TEST_BERUFSBILDUNG_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"BERUFSBILDUNGDATENEXPORTE_TEST_BERUFSBILDUNG_ENTID": idmap,
		"BERUFSBILDUNGDATENEXPORTE_TEST_LIVE":      "FALSE",
		"BERUFSBILDUNGDATENEXPORTE_TEST_EXPLAIN":   "FALSE",
		"BERUFSBILDUNGDATENEXPORTE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["BERUFSBILDUNGDATENEXPORTE_TEST_BERUFSBILDUNG_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["BERUFSBILDUNGDATENEXPORTE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["BERUFSBILDUNGDATENEXPORTE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewBerufsbildungDatenExporteSDK(core.ToMapAny(mergedOpts))
	}

	live := env["BERUFSBILDUNGDATENEXPORTE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["BERUFSBILDUNGDATENEXPORTE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/BerufsbildungEntityTest.php
<?php
declare(strict_types=1);

// Berufsbildung entity test

require_once __DIR__ . '/../berufsbildungdatenexporte_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class BerufsbildungEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = BerufsbildungDatenExporteSDK::test(null, null);
        $ent = $testsdk->Berufsbildung(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "berufsbildung" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = BerufsbildungDatenExporteSDK::test($seed, null);
        $seen = iterator_to_array($base->Berufsbildung(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = BerufsbildungDatenExporteConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = BerufsbildungDatenExporteSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Berufsbildung(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = berufsbildung_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "berufsbildung." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set BERUFSBILDUNG_DATEN_EXPORTE_TEST_BERUFSBILDUNG_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $berufsbildung_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.berufsbildung")));
        $berufsbildung_ref01_data = null;
        if (count($berufsbildung_ref01_data_raw) > 0) {
            $berufsbildung_ref01_data = Helpers::to_map($berufsbildung_ref01_data_raw[0][1]);
        }

        // LIST
        $berufsbildung_ref01_ent = $client->Berufsbildung(null);
        $berufsbildung_ref01_match = [];

        $berufsbildung_ref01_list_result = $berufsbildung_ref01_ent->list($berufsbildung_ref01_match, null);
        $this->assertIsArray($berufsbildung_ref01_list_result);

        // LOAD
        $berufsbildung_ref01_match_dt0 = [];
        $berufsbildung_ref01_data_dt0_loaded = $berufsbildung_ref01_ent->load($berufsbildung_ref01_match_dt0, null);
        $this->assertNotNull($berufsbildung_ref01_data_dt0_loaded);

    }
}

function berufsbildung_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/berufsbildung/BerufsbildungTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = BerufsbildungDatenExporteSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["berufsbildung01", "berufsbildung02", "berufsbildung03", "export01", "export02", "export03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("BERUFSBILDUNG_DATEN_EXPORTE_TEST_BERUFSBILDUNG_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "BERUFSBILDUNG_DATEN_EXPORTE_TEST_BERUFSBILDUNG_ENTID" => $idmap,
        "BERUFSBILDUNG_DATEN_EXPORTE_TEST_LIVE" => "FALSE",
        "BERUFSBILDUNG_DATEN_EXPORTE_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["BERUFSBILDUNG_DATEN_EXPORTE_TEST_BERUFSBILDUNG_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["BERUFSBILDUNG_DATEN_EXPORTE_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new BerufsbildungDatenExporteSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["BERUFSBILDUNG_DATEN_EXPORTE_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["BERUFSBILDUNG_DATEN_EXPORTE_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
